gangplank/ocp/bc.go

```diff
@@ -223,136 +214,277 @@ binary build interface.`)
+						if info == nil || info.IsDir() {
+							return nil
+						}
+						n := filepath.Base(info.Name())
+						if !(strings.HasPrefix(n, "meta") && strings.HasSuffix(n, ".json")) {
+							l.WithField("file", n).Warning("excluded")
+							return nil
+						}
+						key := filepath.Join(keyPathBase, n)
+						ws.RemoteFiles = append(
+							ws.RemoteFiles,
+							&RemoteFile{
+								Bucket: "builds",
+								Minio:  m,
+								Object: key,
+							},
+						)
+						l.WithFields(log.Fields{
+							"file":   info.Name(),
+							"bucket": "builds",
+							"key":    key,
+						}).Info("Included metadata")
 						return nil
+					})
+
+					foundCount := 0
+					for _, artifact := range s.RequireArtifacts {
+						bArtifact, err := mBuild.GetArtifact(artifact)
+						if err != nil {
+							l.WithField("artifact", artifact).Debug("stage artifact dependency not meet")
+							return false
+						}
+
+						// get the Minio relative path for the object
+						// the full path needs to be broken in to <BUILDID>/<ARCH>/<FILE>
+						keyPath := filepath.Join(keyPathBase, filepath.Base(bArtifact.Path))
+						r := &RemoteFile{
+							Artifact: bArtifact,
+							Bucket:   "builds",
+							Minio:    m,
+							Object:   keyPath,
+						}
+						ws.RemoteFiles = append(ws.RemoteFiles, r)
+						foundCount++
 					}
-					keyPath := filepath.Join(buildPath, n)
-					ws.RemoteFiles = append(
-						ws.RemoteFiles,
-						&RemoteFile{
-							Bucket: "builds",
-							Minio:  m,
-							Object: keyPath,
-						},
-					)
-					log.WithField("file", keyPath).Info("Sending meta-data to worker")
-					return nil
-				})
-
-				for _, artifact := range s.RequireArtifacts {
-					bArtifact, err := mBuild.GetArtifact(artifact)
-					if err != nil {
-						log.WithFields(log.Fields{
-							"stage":    s.ID,
-							"artifact": artifact,
-						}).Info("Waiting on artifact")
-						time.Sleep(15 * time.Second)
-						break
+
+					// Once all the dependencies are met, inform the out channel exit.
+					if len(s.RequireArtifacts) == foundCount {
+						out <- true
+						return true
 					}
+					return false
+				}
 
-					// get the Minio relative path for the object
-					// the full path needs to be broken in to <BUILDID>/<ARCH>/<FILE>
-					keyPath := filepath.Join(buildPath, filepath.Base(bArtifact.Path))
-
-					log.WithFields(log.Fields{
-						"artifact path": keyPath,
-						"artifact":      artifact,
-						"buildID":       buildID,
-						"stage":         s.ID,
-					}).Info("Found required artifact")
-
-					r := &RemoteFile{
-						Artifact: bArtifact,
-						Bucket:   "builds",
-						Minio:    m,
-						Object:   keyPath,
+				for {
+					if check() {
+						l.Debug("all dependencies for stage have been meet")
+						break
+					}
+					// Wait for the next check or terminate.
+					select {
+					case <-terminate:
+						return
+					case <-time.After(15 * time.Second):
```

Comment:
(optional) Is there a way we can edge trigger on this?  Like have a "fetched full build" channel?

Reply:
Can you expand on your thinking? I'm not grawking the idea. The select will only be hit in the case of  `len(s.RequiredArtfiacts) > 0`. Are you looking for a "fetch the full build" option? 

I'm completely open to expanding for nuanced-edged conditions.